Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/.gitattributes

```diff
@@ -0,0 +1,10 @@
+# See https://git-scm.com/docs/gitattributes for more about git attribute files.
```

Comment:
what is this file for?

Reply:
`.gitattribute` is used to specify files and pathnames with the attributes that should be used by git when performing certain git actions. Here are some examples:
`*			text=auto` to have all files identified as text by git to have the end of line converted to `LF` i.e. `\n`
`*.txt		text` to have all files ending with `.txt` to be identified as text
`*.vcproj		text eol=crlf` to have all files ending with `.vcproj` to force to have end of line converted to `CRLF` i.e. `\r`
`*.sh		text eol=lf` to have all files ending with `.sh` to force to have end of line converted to `LF` i.e. `\n`
`*.jpg		-text` to have all files ending with `.jpg` not to be normalized as text

Reply:
So, how is this file different than `.gitignore` ?

Reply:
`.gitignore` is used to include file path to those files that you wanted to be ignored during your git push.
`.gitattribute` on the other hand is used to give attributes to files by specifying filenames or pathnames and the given attributes.
Other than those example mentioned above, for example inside this particular `.gitattribute` file generated by Rails.
```
# Mark the database schema as having been generated.
db/schema.rb linguist-generated

# Mark the yarn lockfile as having been generated.
yarn.lock linguist-generated

# Mark any vendored files as having been vendored.
vendor/* linguist-vendored
```
`db/schema.rb` is the path to the specific file while `linguist-generated` identified the specific file as a generated code, same goes to the `yarn.lock` file.
`vendor/*` indicates that all files inside the `vendor` folder will be identified as vendored code. Vendor folder is a directory to hold third party files or codes such as plugins and gems in Rails.
Both `linguist-generated` and `linguist-vendored` indicates that these files will be excluded by Git Linguist when generating the overview of languages contained in the repository.